File: analyzer/src/lexer/token.rs
use super::token_type::TokenType;

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    token_type: TokenType,
    span: Span,
    text: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Span {
    start: Position,
    end: Position,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Position {
    row: usize,
    col: usize,
    cursor_start: usize,
}

impl Token {
    pub fn new(token_type: TokenType, span: Span, text: String) -> Self {
        Self {
            token_type,
            span,
            text,
        }
    }

    pub fn token_type(&self) -> &TokenType {
        &self.token_type
    }

    pub fn span(&self) -> &Span {
        &self.span
    }

    pub fn text(&self) -> &str {
        self.text.as_ref()
    }
}

impl Span {
    pub fn new(start: Position, end: Position) -> Self {
        Self { start, end }
    }

    pub fn start(&self) -> &Position {
        &self.start
    }

    pub fn end(&self) -> &Position {
        &self.end
    }
}

impl Position {
    pub fn new(row: usize, col: usize, cursor_start: usize) -> Self {
        Self {
            row,
            col,
            cursor_start,
        }
    }

    pub fn row(&self) -> usize {
        self.row
    }

    pub fn col(&self) -> usize {
        self.col
    }

    pub fn cursor_start(&self) -> usize {
        self.cursor_start
    }
}

File: src/Lexer/Token/Token.cpp
#include "Lexer/Token/Token.hpp"
#include "Lexer/Lexer.hpp"
#include "Token.hpp"

namespace Marble
{
    Position::Position(size_t row, size_t col, size_t cursor) : Row{row}, Col{col}, Cursor{cursor}
    {
    }

    Span::Span(const Position &start, const Position &end) : Start{start}, End{end}
    {
    }

    Token::Token(Marble::TokenType tokenType, const std::string &text, const Marble::Span &span)
        : m_TokenType{tokenType}, m_Text{text}, m_Span{span}
    {
    }

    Token::Token(Lexer &lexer, Marble::TokenType tokenType)
        : Token{tokenType, lexer.File().Content().substr(lexer.CursorStart(), lexer.Cursor() - lexer.CursorStart()),
                Marble::Span(Position(lexer.Line() + 1, lexer.CursorStart() - lexer.LineStart(), lexer.CursorStart()),
                             Position(lexer.Line() + 1, lexer.Cursor() - lexer.LineStart(), lexer.Cursor()))}
    {
    }

    Token::Token(const Token &token) : Token{token.m_TokenType, token.m_Text, token.m_Span}
    {
    }

} // namespace Marble
